migrations/lib/importer/steps/base/site_settings.rb

```diff
@@ -0,0 +1,231 @@
+# frozen_string_literal: true
+
+module Migrations::Importer::Steps::Base
+  class SiteSettings < ::Migrations::Importer::Step
+    class << self
+      def inherited(klass)
+        super
+        klass.requires_mapping :last_change_dates, "SELECT name, updated_at FROM site_settings"
+      end
+    end
+
+    DATATYPES_WITH_DEPENDENCY =
+      [
+        Enums::SiteSettingDatatype::CATEGORY,
+        Enums::SiteSettingDatatype::CATEGORY_LIST,
+        Enums::SiteSettingDatatype::EMOJI_LIST,
+        Enums::SiteSettingDatatype::GROUP,
+        Enums::SiteSettingDatatype::GROUP_LIST,
+        Enums::SiteSettingDatatype::TAG_GROUP_LIST,
+        Enums::SiteSettingDatatype::TAG_LIST,
+        Enums::SiteSettingDatatype::UPLOAD,
+        Enums::SiteSettingDatatype::UPLOADED_IMAGE_LIST,
+        Enums::SiteSettingDatatype::USERNAME,
+      ].map { |number| ::SiteSettings::TypeSupervisor.types[number].to_s }
+
+    DISALLOWED_SITE_SETTINGS = Set.new([:permalink_normalizations]).freeze
+    LIST_SEPARATOR = "|"
+
+    private_constant :DATATYPES_WITH_DEPENDENCY, :DISALLOWED_SITE_SETTINGS
+
+    def execute
+      super
+
+      @log_messages = []
+      @audit_message = I18n.t("importer.site_setting_log_message")
+      @failed_settings = []
+
+      @settings_index =
+        silence_deprecation_warnings do
+          SiteSetting.all_settings(include_hidden: true)
+        end.index_by { |hash| hash[:setting] }
+
+      import_rows = fetch_rows.reject { |row| skip_row?(row) }
+
+      with_progressbar(import_rows.size) do
+        capture_deprecation_warnings { process_rows(import_rows) }
+      end
+
+      # TODO: Replace with logging when available
+      @log_messages.each { |message| puts "      #{message}" }
+    end
+
+    private
+
+    # Override in subclasses to skip rows that should be handled in a later step.
+    def skip_row?(_row)
+      false
+    end
+
+    # Override in subclasses to adapt/normalize values by type.
+    def transform_value(_setting_name, value, _type)
+      value
+    end
+
+    private
+
+    def fetch_rows
+      @intermediate_db.query <<~SQL
+        SELECT name, value, last_changed_at, import_mode
+        FROM site_settings
+        ORDER BY ROWID
+      SQL
+    end
+
+    def process_rows(rows)
+      rows.each do |row|
+        @stats.reset
+        before_fail_count = @failed_settings.size
+
+        import_row(row) if row_importable?(row)
+
+        # Only advance the progress bar if the row didn't get deferred due to failure.
+        update_progressbar if before_fail_count == @failed_settings.size
+      end
+
+      # Re-attempt failed updates (often due to dependency ordering).
```

Comment:
Discourse doesn't have a concept of dependencies between settings. All we can do is re-attempt to change a setting multiple times. Maybe it works after another setting was changed as well. If all attempts fail, we log it.